File: pdf-rag-ts/src/generation/generator.ts
export class AnswerGenerator {
    private ollamaUrl: string;
    private generativeModel: string;

    constructor(ollamaUrl: string, generativeModel: string) {
        this.ollamaUrl = ollamaUrl;
        this.generativeModel = generativeModel;
    }

    async generate(prompt: string): Promise<string> {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), 120000);

        try {
            const response = await fetch(`${this.ollamaUrl}/api/generate`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    model: this.generativeModel,
                    prompt: prompt,
                    stream: false,
                }),
                signal: controller.signal,
            });

            clearTimeout(timeoutId);

            if (!response.ok) {
                throw new Error(`Ollama error: ${response.statusText}`);
            }

            const data = await response.json() as { response?: string };
            return (data.response || "").trim();
        } catch (error) {
            clearTimeout(timeoutId);
            throw error;
        }
    }
}

File: pdf-rag-python/src/generation/generator.py
import requests


class AnswerGenerator:
    def __init__(self, ollama_url: str, generative_model: str):
        self.ollama_url = ollama_url
        self.generative_model = generative_model

    def generate(self, prompt: str) -> str:
        response = requests.post(
            f"{self.ollama_url}/api/generate",
            json={"model": self.generative_model, "prompt": prompt, "stream": False},
            timeout=120,
        )
        response.raise_for_status()
        return response.json().get("response", "").strip()
